flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java

```diff
@@ -270,7 +270,9 @@ private InputStatus finishedOrAvailableLater() {
 			splitFetcherManager.checkErrors();
 			return InputStatus.END_OF_INPUT;
 		} else {
-			throw new IllegalStateException("Called 'finishedOrAvailableLater()' with shut-down fetchers but non-empty queue");
+			// We may fall in this method because of purely finished splits record. If all splits
```

Comment:
Maybe change this to:

> We can reach this case if we just processed all data from the queue and finished a split, and concurrently the fetcher finished another split, whose data is then in the queue.

Reply:
@StephanEwen I have changed the comment to what you recommend in fixup commit. In addition to this, I add another fixup commit with new test case `testMultipleSplitsWithDifferentFinishingMoments` which I think is more common than `testMultipleSplitsWithSeparatedFinishedRecord`. After changes, I think the comment covers general concurrent problem while not tied to a specific case, in my opinion.

Both above fixup commits should be squashed out before merging.